File: RKExtensions/RKExtensions-ObjC/Classes/Controller/RKNavigationController.m
//
//  RKNavigationController.m
//  RKExtensions-ObjC
//
//  Created by 李沛倬 on 2018/6/3.
//  Copyright © 2018年 peizhuo. All rights reserved.
//

#import "RKNavigationController.h"

@interface RKNavigationController ()<UIGestureRecognizerDelegate, UINavigationControllerDelegate>

@end

@implementation RKNavigationController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.delegate = self;

    [self replacePopGesture];
    
}

- (void)replacePopGesture {
    
    SEL sel = NSSelectorFromString(@"handleNavigationTransition:");
    
    if ([self.interactivePopGestureRecognizer.delegate respondsToSelector:sel]) {
        
        UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self.interactivePopGestureRecognizer.delegate action:sel];
        
        pan.delegate = self;
        
        [self.view addGestureRecognizer:pan];
    }
    
}


// MARK: - Action

- (void)backBarItemAction:(UIBarButtonItem *)sender {
    
    [self popViewControllerAnimated:true];
}


// MARK: - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    
    if ([self respondsToSelector:@selector(prefersFullscreenPopGesture)]) {
        return [self prefersFullscreenPopGesture];
    }
    
    return self.viewControllers.count != 1;
}


// MARK: - UINavigationControllerDelegate

- (void)navigationController:(UINavigationController *)navigationController willShowViewController:(UIViewController *)viewController animated:(BOOL)animated {
    
    [self configNavigationBarBackItem];
    
}

- (void)navigationController:(UINavigationController *)navigationController didShowViewController:(UIViewController *)viewController animated:(BOOL)animated {
    
}



// MARK: - Private Method

- (void)configNavigationBarBackItem {
    if (self.viewControllers.count < 2 || !self.backItemImage) { return; }
    
    UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithImage:self.backItemImage style:UIBarButtonItemStylePlain target:self action:@selector(backBarItemAction:)];
    
    self.topViewController.navigationItem.leftBarButtonItem = backItem;
}




@end

File: RKExtensions/RKExtensions-Swift/Classes/Controller/RKNavigationController.swift
//
//  RKNavigationController.swift
//  RKExtensions-Swift
//
//  Created by 李沛倬 on 2018/7/12.
//  Copyright © 2018年 peizhuo. All rights reserved.
//

import UIKit

class RKNavigationController: UINavigationController {
    
    public var backItemImage : UIImage?

    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.delegate = self
        self.replacePopGesture()
    }
    
    fileprivate func replacePopGesture() {
        
        let sel = NSSelectorFromString("handleNavigationTransition:")
        
        if self.interactivePopGestureRecognizer?.delegate?.responds(to: sel) == true {
            let pan = UIPanGestureRecognizer.init(target: self.interactivePopGestureRecognizer?.delegate, action: sel)
            pan.delegate = self
            
            self.view.addGestureRecognizer(pan)
        }
    }
    
    
    fileprivate func configNavigationBarBackItem() {
        if self.viewControllers.count < 2 || self.backItemImage == nil { return }
        
        let backItem = UIBarButtonItem.init(image: self.backItemImage!, style: .plain, target: self, action: #selector(backBarItemAction(sender:)))
        
        self.topViewController?.navigationItem.leftBarButtonItem = backItem
    }

    
    @objc fileprivate func backBarItemAction(sender : UIBarButtonItem) {
        self.popViewController(animated: true)
    }

}


// MARK: - UIGestureRecognizerDelegate

extension RKNavigationController : UIGestureRecognizerDelegate {
    
    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        
        
        return self.viewControllers.count != 1
    }
}


// MARK: - UINavigationControllerDelegate

extension RKNavigationController : UINavigationControllerDelegate {
    
    func navigationController(_ navigationController: UINavigationController, willShow viewController: UIViewController, animated: Bool) {
        
        self.configNavigationBarBackItem()
        
        print_debug(viewController)
    }

    
}































